### DISCUSSION TITLE
K3s regularly spikes in cpu usage and crashes in a loop

### DISCUSSION THREAD

**Discussion by Blackclaws** [k3s-io/k3s#7786] [General]
**Environmental Info:**
K3s Version: 
k3s version v1.24.3+k3s1 (990ba0e8)
go version go1.18.1

Node(s) CPU architecture, OS, and Version: 
Almalinux 8.6, x86_64 8 Cores

Cluster Configuration:
1 Server in HA ready configuration

**Describe the bug:**
k3s regularly spikes into 700-800% cpu usage until it crashes because the leaderelection is lost (to itself). It then reboots and repeats this loop. 

This happens about every 2-3 days.

**Additional context / logs:**

Manually stopping k3s with systemctl stop k3s and restarting it, fixes it temporarily.

According to the logs etcd apply takes up to 6 seconds sometimes.
```
Sep 02 15:13:22 miku k3s[2694852]: I0902 15:13:22.250924 2694852 trace.go:205] Trace[1434124888]: "List(recursive=true) etcd3" key:/events,resourceVersion:0,resourceVersionMatch:,limit:500,continue: (02-Sep-2022 15:13:14.773) (total time: 7477ms):
Sep 02 15:13:22 miku k3s[2694852]: Trace[1434124888]: [7.477022009s] [7.477022009s] END
```
```
Sep 02 15:13:22 miku k3s[2694852]: E0902 15:13:22.349266 2694852 leaderelection.go:330] error retrieving resource lock kube-system/kube-controller-manager: Get "https://127.0.0.1:6444/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-controller-manager?timeout=5s": context deadline exceeded
Sep 02 15:13:22 miku k3s[2694852]: I0902 15:13:22.349361 2694852 leaderelection.go:283] failed to renew lease kube-system/kube-controller-manager: timed out waiting for the condition
Sep 02 15:13:22 miku k3s[2694852]: E0902 15:13:22.349426 2694852 controllermanager.go:284] "leaderelection lost"
```

**Comment by brandond** [comment]
That usually indicates that the storage isn't sufficient to keep up with the IO requirements of the datastore. By "HA ready configuration", I assume you mean a single-node etcd cluster? Etcd has much higher IO requirements than sqlite, and really won't work well on anything except for SSD. What sort of disk are you running k3s on? Are you using the same physical media for the datastore, your workload, and the containerd image store?

**Comment by Blackclaws** [comment]
The thing is that etcd provides native snapshotting and this is a server we might want to expand by worker nodes in the future.

The underlying storage are NVME ssds that _should_ have sufficient headroom for fast access. The same physical medium is used for datastore, workload and containerd images as this is a VM that only has a single volume provisioned.

What's surprising to me is that during initial startup this isn't a problem at all, but only seems to become a problem once some time has passed. Is there any way to increase the resilience to these effects?

**Comment by brandond** [comment]
> The underlying storage are NVME ssds that should have sufficient headroom for fast access. The same physical medium is used for datastore, workload and containerd images as this is a VM that only has a single volume provisioned.

Have you profiled the disk for latency and throughput? Keep in mind that etcd makes a blocking `fsync` call for **every** write operation to the datastore. If your workload is performing a lot of write operations at the same time, all of those will need to be flushed to disk before etcd can commit the write. You may not be exceeding the rated iops, but simply running into problems with bulk IO blocking performance-critical cluster operations.

> What's surprising to me is that during initial startup this isn't a problem at all, but only seems to become a problem once some time has passed.

Is this perhaps occurring during some background operation, such as a scheduled `TRIM` operation, or some IO-intensive processing job on the part of the workload?

**Comment by Blackclaws** [comment]
> > The underlying storage are NVME ssds that should have sufficient headroom for fast access. The same physical medium is used for datastore, workload and containerd images as this is a VM that only has a single volume provisioned.
> 
> Have you profiled the disk for latency and throughput? Keep in mind that etcd makes a blocking `fsync` call for **every** write operation to the datastore. If your workload is performing a lot of write operations at the same time, all of those will need to be flushed to disk before etcd can commit the write. You may not be exceeding the rated iops, but simply running into problems with bulk IO blocking performance-critical cluster operations.
> 

Yes. Throughput and latency should be more than sufficient.

> > What's surprising to me is that during initial startup this isn't a problem at all, but only seems to become a problem once some time has passed.
> 
> Is this perhaps occurring during some background operation, such as a scheduled `TRIM` operation, or some IO-intensive processing job on the part of the workload?

No, it happens _seemingly_ at random. Manually running trim operations doesn't trigger this, writing a lot of data doesn't trigger this. I honestly don't know _what_ triggers this. iostat while it happens shows the storage not particularly under pressure, CPU usage at full all taken by a multitude of k3s processes.

Any sort of process that I might follow to help me debug the cause of this?

**Comment by brandond** [comment]
> CPU usage at full all taken by a multitude of k3s processes.

Can you clarify what you mean by that? There should normally only be 1 k3s process (either `k3s server` or `k3s agent`), occasionally 2 if k3s is started with the --log flag and has to re-exec itself to redirect output to a log file. Are you actually seeing multiple k3s processes, or just multiple OS threads?

**Comment by Blackclaws** [comment]
Multiple OS Threads I'd say. I've attached a view of htop and iostat of the problem occuring. As I said iostat is rather unproblematic. IoWait is low, but for some reason k3s uses a ton of cpu and times out etcd requests:

![k3s-usage](https://user-images.githubusercontent.com/5792929/192353678-708ee670-6515-4cd1-a382-f33fba7c2f59.png)
![k3s-iostat](https://user-images.githubusercontent.com/5792929/192353939-383b5d85-c34f-4aa8-802d-b4d755da78a1.png)

**Comment by brandond** [comment]
If you're using iostat, can you look at the `f_await` and `aqu-sz` fields? These are more likely to be what's causing etcd problems than just the base IO rate.

https://manpages.debian.org/testing/sysstat/iostat.1.en.html#f_await

**Comment by Blackclaws** [comment]
Will do the next time this problem shows up. Since this is a production server restoring service had priority over waiting to debug this.

**Comment by Blackclaws** [comment]
So aqu-sz is really low, f_await isn't tracked, this might be because the underlying disk is a virtual disk.

What I've also noticed is a spike from ```iptables-save -t filter``` that saves current filter rules, I think that's part of the startup/shutdown of k3s. 

Just to give you an idea of the thread sprawl that k3s seems to produce in this situation:

![Screenshot_2022-11-07_09-17-12](https://user-images.githubusercontent.com/5792929/200259843-36f71a08-b7a4-41e7-b5ce-becb80300091.png)

The machine basically becomes unusable at this point. IO latency does also spike at this point (if I run a benchmark like bonnie++ at the same time) however its hard to tell whether that is simply because of the high load on the machine. Shutting down k3s returns it to normal after which k3s can be restarted and normal operation resumes.

I won't rule out that this is an issue that crops up due to the underlying machine. What I find weird is that it takes quite some time to happen the first time, even with the full load of cluster startup directly applied at the beginning (which I assume is the point of highest load for most use cases). There seems to be some sort of trigger that puts k3s in this state where a recovery takes a:
```
k3s-killall.sh
systemctl start k3s
```
Which then leads to a standard startup that works fine.

I wonder why the cpu load spikes this way however, it seems there is work being done since the processes aren't IO blocked. I wonder _what_ work is being done however.

**Comment by brandond** [comment]
I haven't seen any updates on this in a bit; were you able to identify a root cause of the CPU load?

**Comment by Blackclaws** [comment]
@brandond Unfortunately not. The issue persists and crops up every once in a while. I can't rule out that it has _something_ to do with disk io being too slow, however we've switched VM hosts etc. It might be related to other VMs running on the same host. Since we don't control the host directly its hard to tell whether the problems coincide with high IO load on the host system. 

However the fact that this seems to crop up semi-regularly and the only way to make it stop is to reboot the machine (after which the problem goes away) and seeing how a simple restart of k3s _without_ reboot does nothing to alleviate the issue it might be related to some sort of failure in the IO system that gets cleared. 

Its still extremely weird and I'd love to go a second round on this if you have any further ideas on what to test. Especially since the very high CPU load is a bit weird. It basically drives the whole system into the ground because k3s takes up so much of the resources that the existing containers start to stall (followed by k3s itself stalling and timing out, restarting and repeating the cycle). I get the feeling that since k3s is supposed to run on even low end systems this is not something that _should_ be happening on this type of host even using the etcd backend.

**Comment by caroline-suse-rancher** [comment]
Converting this to a discussion as (at this point), it doesn't seem to be directly related to a k3s bug.

**Comment by PatrickHuetter** [comment]
Hey together, anything new on this? Having the same problem with k3s on a VM. Storage IO should be sufficient. Is it possible that this has something to do with the iptables binary which gets used? 

Current setup:
k3s v1.31.2+k3s1 using wireguard-native installed via official k3s ansible playbook
Prepared for HA, but currently running as single instance node
IPv4/IPv6 dual stack activated
tailscale installed
iptables in use (firewall)
4 Cores, 4G RAM, Ubuntu 24.04.1 LTS (GNU/Linux 6.8.0-51-generic x86_64)

![Bildschirmfoto_2025-01-11_11-40-25](https://github.com/user-attachments/assets/70055ccb-e8f1-4d9f-a88a-802bd0db1afd)

**Comment by PatrickHuetter** [comment]
@brandond Just to let you know. If i'm not wrong, the "HA-ready" configuration in single node operation doesn't use etcd. It uses sqlite and switches to etcd if you add another node. If you don't add another node you stay at sqlite automatically. Please corrent me if i'm wrong. So i think the problem isn't etcd related, it's sqlite related. I could see that the sqlite state.db file grows to many GB. Mine was 10GB after 3 weeks of uptime (new k3s node). I already tested some "vacuum" tools for sqlite, that did some shrinking on the db but it didn't solve the problem with the high cpu load.

**Reply by dereknola** [reply]
That is not correct. If you start a node with `k3s server --cluster-init`, it starts with an embedded etcd DB.
Alternatively, if you
1) Start a server with no cluster-init `k3s server`, you get a single node server with sqlite DB
2) Try and join a second server to that first one `k3s server --server=https://SERVER-1.IP:6443`
The second server will fail to join, no automatic conversion will happen. We require the user to intervene on purpose to migrate from sqlite to etcd https://docs.k3s.io/datastore/ha-embedded#existing-single-node-clusters
